code/variant/variant.cpp

```diff
@@ -28,6 +30,12 @@ struct Neutron : public Particle {
   virtual void print() const { std::cout << "N" << std::endl; }
 };
 
+std::size_t countNeutrons(std::span<const std::unique_ptr<Particle>> ps) {
+    return std::count_if(ps.begin(), ps.end(), [](const std::unique_ptr<Particle>& p){
+        return typeid(*p) == typeid(Neutron);
```

Comment:
Would dynamic_cast be better here over RTTI even for showing problematic code, just to prevent someone from actually abusing this IRL ;)

Reply:
Hmm, I would implement it with `typeid`, which is less of a hammer than `dynamic_cast`. But AFAIK we also do not discuss `typeid` in the lecture, so maybe `dynamic_cast` is the better choice than. And sure, it is scary!

Reply:
If the intent is to intriduce a better use-case for `get_if`, I think `dynamic_cast` is more suited, because having a similar mechanism, which is to return a null pointer if the expected type is not the real one. Also, the spirit of the exercice being the move from an "OO paradigm" to a "Functional paradigm", I think `dynamic_cast` is a lot more "typical OO idiom". Yes, it is very subjective ;)

Reply:
Since we are moving down an inheritance hierarchy rather than 2 disjoint classes, I kind of feel like `dynamic_cast` is a better alternative like @chavid  mentions. Also the rare times you've had to inspect types at runtime a common case is when you'd have to invoke some specific method from the derived class in which case you're bound to `dynamic_cast` (which is ofc a bad design... but life;) 
Not a physicist, but for this example you could add something like a `get_antiparticle` which returns a class and since one can't do this anymore with inheritance you'd have the some derived classes implementing this or not.

This would also be a good use case for using variant of particles as it shows the use case for the times you'd want to unify types that don't need a hierarchy ;)